File: programmers/lv3/합승 택시 요금.py
from collections import deque, defaultdict

def solution(n, s, a, b, fares):
    answer, fare_dict = 0, defaultdict(list)
    
    for fare in fares:
        fare_dict[fare[0]].append([fare[1], fare[2]])
        fare_dict[fare[1]].append([fare[0], fare[2]])
    
    nodes = [float('inf')] * (n + 1)
    nodes[s], q = 0, deque([s])
    
    while q:
        node = q.popleft()
        for n_node, cost in fare_dict[node]:
            if nodes[n_node] > nodes[node] + cost:
                nodes[n_node] = nodes[node] + cost
                q.append(n_node)
    
    answer = nodes[a] + nodes[b]
    
    for i in range(1, n + 1):
        if i == s: continue
        new_nodes = [float('inf')] * (n + 1)
        new_nodes[i], q = nodes[i], deque([i])
        while q:
            node = q.popleft()
            for n_node, cost in fare_dict[node]:
                if new_nodes[n_node] > new_nodes[node] + cost:
                    new_nodes[n_node] = new_nodes[node] + cost
                    q.append(n_node)
        answer = min(answer, new_nodes[a] + new_nodes[b] - new_nodes[i])
    
    return answer

File: programmers/lv3/합승 택시 요금.java
import java.util.*;

class Solution {
    public int solution(int n, int s, int a, int b, int[][] fares) {
        int answer = Integer.MAX_VALUE;
        
        int[][] minCosts = new int[n + 1][n + 1];
        for(int i = 1; i <= n; i++) {
            for(int j = 1; j <= n; j++) {
                minCosts[i][j] = i == j ? 0 : Integer.MAX_VALUE;
            }
        }
        
        for(int[] fare: fares) {
            int e1 = fare[0], e2 = fare[1], cost = fare[2];
            minCosts[e1][e2] = cost;
            minCosts[e2][e1] = cost;
        }
        
        
        for(int k = 1; k <= n; k++) {
            for(int i = 1; i <= n; i++) {
                for(int j = 1; j <= n; j++) {
                    if(minCosts[i][k] == Integer.MAX_VALUE || minCosts[k][j] == Integer.MAX_VALUE) {
                        continue;
                    }
                    minCosts[i][j] = Math.min(minCosts[i][j], minCosts[i][k] + minCosts[k][j]);
                }
            }
        }
        
        for(int start = 1; start <= n; start++) {
            int[] row = minCosts[start];
            if(row[a] != Integer.MAX_VALUE && row[b] != Integer.MAX_VALUE && row[s] != Integer.MAX_VALUE) {
                answer = Math.min(answer, row[a] + row[b] + row[s]);
            }
        }
        
        return answer;
    }
}
